crates/swift-bridge-ir/src/codegen/codegen_tests/async_function_codegen_tests.rs

```diff
@@ -533,3 +533,107 @@ void __swift_bridge__$SomeType$some_method(void* callback_wrapper, void __swift_
         .test();
     }
 }
+
+mod extern_rust_async_function_returns_result_opaque {
+    use super::*;
+
+    fn bridge_module() -> TokenStream {
+        quote! {
+            #[swift_bridge::bridge]
+            mod ffi {
+                extern "Rust" {
+                    type SomeType1;
+                    type SomeType2;
+                    async fn some_function() -> Result<SomeType1, SomeType2>;
+                }
+            }
+        }
+    }
+
+    fn expected_rust_tokens() -> ExpectedRustTokens {
+        ExpectedRustTokens::Contains(quote! {
+             pub extern "C" fn __swift_bridge__some_function(
+                callback_wrapper: *mut std::ffi::c_void,
+                callback: extern "C" fn(*mut std::ffi::c_void, swift_bridge :: result :: ResultPtrAndPtr) -> (),
+            ) {
+                let callback_wrapper = swift_bridge::async_support::SwiftCallbackWrapper(callback_wrapper);
+                let fut = super::some_function();
+                let task = async move {
+                let val = match fut.await {
+                    Ok(ok) => {
+                        swift_bridge::result::ResultPtrAndPtr {
+                            is_ok: true,
+                            ok_or_err: Box::into_raw(Box::new(ok)) as *mut super::SomeType1 as *mut std::ffi::c_void
+                        }
+                    }
+                    Err(err) => {
+                        swift_bridge::result::ResultPtrAndPtr {
+                            is_ok: false,
+                            ok_or_err: Box::into_raw(Box::new(err)) as *mut super::SomeType2 as *mut std::ffi::c_void
+                        }
+                    }
+                };
+                    let callback_wrapper = callback_wrapper;
+                    let callback_wrapper = callback_wrapper.0;
+
+                    (callback)(callback_wrapper, val)
+                };
+                swift_bridge::async_support::ASYNC_RUNTIME.spawn_task(Box::pin(task))
+            }
+        })
+    }
+
+    fn expected_swift_code() -> ExpectedSwiftCode {
+        ExpectedSwiftCode::ContainsAfterTrim(
+            r#"
+public func some_function() async throws -> SomeType1 {
+    func onComplete(cbWrapperPtr: UnsafeMutableRawPointer?, rustFnRetVal: __private__ResultPtrAndPtr) {
+        let wrapper = Unmanaged<CbWrapper$some_function>.fromOpaque(cbWrapperPtr!).takeRetainedValue()
+        if rustFnRetVal.is_ok {
+            wrapper.cb(.success(SomeType1(ptr: rustFnRetVal.ok_or_err!)))
+        } else {
+            wrapper.cb(.failure(SomeType2(ptr: rustFnRetVal.ok_or_err!)))
+        }
+    }
+
+    return try await withCheckedThrowingContinuation({ (continuation: CheckedContinuation<SomeType1, Error>) in
```

Comment:
I tried to do this, I got the following error.

```
Testing failed:
	Cannot convert value of type '(CheckedContinuation<AsyncResultOpaqueRustType1, AsyncResultOpaqueRustType2>) -> Void' to expected argument type '(CheckedContinuation<AsyncResultOpaqueRustType1, any Error>) -> Void'
	Testing cancelled because the build failed.

```

Reply:
I made the minimal code to reproduce this error.

```Swift
struct ZeroError: Error{}

func hoge(value: Int) async throws -> Int {
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Int, ZeroError>) in
        if value == 0{
            continuation.resume(throwing: ZeroError())
        }else{
            continuation.resume(returning: value*2)
        }
    }
}
```

Reply:
Thanks for the minimal example.

This compiles:

```swift
struct ZeroError: Error{}

func hoge(value: Int) async throws -> Int {
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Int, _>) in
        if value == 0{
            continuation.resume(throwing: ZeroError())
        }else{
            continuation.resume(returning: value*2)
        }
    }
}
```

Since `ZeroError` doesn't compile leaving it as `Error` is fine.

Would just leave a comment somewhere in the test.

Maybe:

```
TODO: Replace `Error` with the concrete error type `SomeType2`. 
 As of Feb 2023 using the concrete error type leads to a compile time error.
 This seems like a bug in the Swift compiler.
```